src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -156,7 +156,123 @@ bool DiscoveryDataBase::process_data_queue()
+                        }
+                    }
+                }
+                else
+                {
+                    // Add DATA(p) of the client with the writer to `pdp_to_send_` (if it's not there).
+                    if (std::find(
+                        pdp_to_send_.begin(),
+                        pdp_to_send_.end(),
+                        parts_writer_it->second.change()) == pdp_to_send_.end())
+                    {
+                        pdp_to_send_.push_back(parts_writer_it->second.change());
+                    }
+                    // Set topic as not-clearable.
+                    is_clearable = false;
+                }
+
+                // Check in `participants_` whether the client with the writer has acknowledge the PDP of the client
+                // with the reader.
+                if (parts_writer_it != participants_.end() && parts_writer_it->second.is_matched(reader.guidPrefix))
+                {
+                    // Check the status of the reader in `writers_[writer]::relevant_participants_builtin_ack_status`.
+                    auto writers_it = writers_.find(writer);
+                    if (writers_it != writers_.end() && !writers_it->second.is_matched(reader.guidPrefix))
+                    {
+                        // If the status is 0, add DATA(r) to a `edp_subscriptions_to_send_` (if it's not there).
+                        if (std::find(
+                            edp_subscriptions_to_send_.begin(),
+                            edp_subscriptions_to_send_.end(),
+                            readers_it->second.change()) == edp_subscriptions_to_send_.end())
+                        {
+                            edp_subscriptions_to_send_.push_back(readers_it->second.change());
+                        }
+                    }
+                }
+                else
+                {
+                    // Add DATA(p) of the client with the reader to `pdp_to_send_` (if it's not there).
+                    if (std::find(
+                        pdp_to_send_.begin(),
+                        pdp_to_send_.end(),
+                        parts_reader_it->second.change()) == pdp_to_send_.end())
+                    {
+                        pdp_to_send_.push_back(parts_reader_it->second.change());
+                    }
+                    // Set topic as not-clearable.
+                    is_clearable = false;
+                }
+            }
+        }
+
+        // Check whether the topic is still dirty or it can be cleared
+        if (is_clearable)
+        {
+            // Delete topic from dirty_topics_
+            topic_it = dirty_topics_.erase(topic_it);
+        }
+        else
+        {
+            // Mark are_dirty_topics as true since this topic is still dirty
+            are_dirty_topics = true;
+            // Proceed with next topic
+            ++topic_it;
+        }
+    }
+    return are_dirty_topics;
```

Comment:
```suggestion
    return dirty_topics_.empty();
```
you can avoid using an extra boolean

Reply:
Good catch :robot: